# Principios de Informática: Tipos de Datos y Variables en Python 🧮
### Componentes fundamentales de todo programa

**Curso:** Principios de Informática

---

## 🗺️ Objetivos y contenidos

Este notebook presenta de forma estructurada los conceptos esenciales de los tipos de datos y el manejo de variables en Python. El propósito es reconocer, utilizar y transformar los tipos de datos más comunes, crear variables con criterios correctos y evitar errores frecuentes.

> Comprender los tipos de datos es un requisito básico para diseñar programas correctos y eficientes.

**Importancia:**
- Los tipos de datos y las variables constituyen la base de cualquier programa.
- Un uso adecuado mejora la claridad, seguridad y eficiencia del código.
- Su dominio permite resolver problemas en contextos reales (cálculo científico, aplicaciones, etc.).

**Contenidos:**
1. Tipos de datos fundamentales: enteros, flotantes, booleanos, cadenas de caracteres
2. Variables: nomenclatura, inicialización, asignación
3. Conversión de tipos
4. Mutabilidad e inmutabilidad

---

## Preámbulo: anotaciones de tipo (type hints) en Python ✍️

Las anotaciones de tipo permiten documentar el tipo esperado de variables y funciones. No cambian la ejecución del programa, pero mejoran la legibilidad, el autocompletado del editor y el análisis estático (p. ej., mypy/pyright).

- Variables: `contador: int = 0`, `promedio: float = 3.5`, `nombre: str = "Ana"`, `activo: bool = True`.
- Colecciones (Python 3.9+): `medidas: list[float]`, `mapa: dict[str, int]`, `ids: set[int]`.
- Uniones: `edad: int | None = None` (equivalente a `Optional[int]`).
- Funciones: `def area(w: float, h: float) -> float: ...`.

Nota: Las anotaciones sirven como contrato de intención; verificación de tipos ocurre con herramientas externas, no en tiempo de ejecución.

---

## 1. Tipos de datos fundamentales

---

Los tipos de datos son las unidades básicas de representación de información en un programa. La selección adecuada permite modelar con precisión y operar de manera eficiente.

> Cada tipo de dato responde a un propósito específico dentro del programa.

En Python, los más utilizados son:

- **Enteros (`int`)**: Números sin parte decimal. Apropiados para conteos exactos (personas, páginas, piezas).
- **Flotantes (`float`)**: Números con parte decimal. Adecuados para mediciones (altura, peso, temperatura).
- **Booleanos (`bool`)**: Valores lógicos con dos estados posibles: `True` o `False`.
- **Cadenas de caracteres (`str`)**: Texto (nombres, mensajes, identificadores, etc.).

A continuación se presentan definiciones y ejemplos de uso de cada tipo.

---

### 🔢 Enteros (`int`)

Los **enteros** representan números sin parte decimal. Son idóneos para conteos, índices y cantidades discretas.

#### Características principales:
- Sin decimales (ej.: 3, -7, 0)
- Positivos, negativos o cero
- Adecuados para conteos exactos y numeraciones
- En Python 3, el tamaño está limitado por la memoria disponible

#### Ejemplos de uso:
- Número de estudiantes en una clase
- Páginas de un libro
- Cantidad de elementos en un inventario
- Año en curso

#### Ejemplo en Python:
```python
edad = 20
cantidad = 5
temperatura_baja = -3
anio_actual = 2025
```

---

In [2]:
# Ejemplo de declaración de variables enteras
revoluciones_motor: int = 5000
poblacion_bacterias: int = 1_000_000_000  # Guiones bajos opcionales para legibilidad
numero_negativo: int = -42
cero: int = 0
print(f"Revoluciones del motor: {revoluciones_motor}")

Revoluciones del motor: 5000


In [3]:
# Imprimir otros ejemplos de enteros
print(f"Población de bacterias: {poblacion_bacterias}")
print(f"Número negativo: {numero_negativo}")
print(f"Cero: {cero}")

Población de bacterias: 1000000000
Número negativo: -42
Cero: 0


In [4]:
# Verificar el tipo de las variables enteras
print(f"Tipo de revoluciones_motor: {type(revoluciones_motor)}")
print(f"Tipo de poblacion_bacterias: {type(poblacion_bacterias)}")

Tipo de revoluciones_motor: <class 'int'>
Tipo de poblacion_bacterias: <class 'int'>


**🧪 Ejercicio: Uso de enteros**

Modificar los valores de las variables enteras, ejecutar el código y analizar los resultados obtenidos.

In [5]:
pulsos_por_segundo: int = 1000
tiempo_segundos: int = 30

# Cálculos automáticos
total_pulsos: int = pulsos_por_segundo * tiempo_segundos
pulsos_por_minuto: int = pulsos_por_segundo * 60

print(f"⚡ Análisis de Pulsos del Sensor")
print(f"Pulsos por segundo: {pulsos_por_segundo}")
print(f"Tiempo de medición: {tiempo_segundos} segundos")
print(f"Total de pulsos detectados: {total_pulsos}")
print(f"Frecuencia por minuto: {pulsos_por_minuto} pulsos/min")

# Verifica que son enteros
print(f"\n🔍 Verificación de tipos:")
for var_name, var_value in [("pulsos_por_segundo", pulsos_por_segundo), 
                           ("total_pulsos", total_pulsos)]:
    print(f"{var_name}: {type(var_value).__name__}")

⚡ Análisis de Pulsos del Sensor
Pulsos por segundo: 1000
Tiempo de medición: 30 segundos
Total de pulsos detectados: 30000
Frecuencia por minuto: 60000 pulsos/min

🔍 Verificación de tipos:
pulsos_por_segundo: int
total_pulsos: int


### 🌊 Flotantes (`float`)

Los **flotantes** representan números reales con parte decimal. Se emplean en mediciones y cantidades fraccionarias.

#### Características principales:
- Incluyen parte decimal (ej.: 3.14, -0.5, 0.0)
- Precisión aproximada (alrededor de 15–17 dígitos significativos)
- Permiten notación científica (ej.: 1.5e-10)
- Apropiados para magnitudes continuas (distancias, temperaturas)

#### Ejemplos de uso:
- Altura de una persona (1.75 m)
- Temperatura ambiente (23.6 °C)
- Peso de un objeto (0.150 kg)
- Distancias astronómicas (1.496e+8 km)

#### Ejemplo en Python:
```python
altura = 1.75
pi = 3.14159
temperatura = 36.5
peso = 0.150
```

---

In [6]:
# Experimentando con flotantes
voltaje: float = 4.95
temperatura: float = 23.7
pi_aproximado: float = 3.14159
constante_planck: float = 6.626e-34  # Notación científica
print(f"Voltaje del sensor: {voltaje} V")

Voltaje del sensor: 4.95 V


In [7]:
# Imprimir otros ejemplos de flotantes
print(f"Temperatura ambiente: {temperatura} °C")
print(f"Pi aproximado: {pi_aproximado}")
print(f"Constante de Planck: {constante_planck} J⋅s")

Temperatura ambiente: 23.7 °C
Pi aproximado: 3.14159
Constante de Planck: 6.626e-34 J⋅s


In [8]:
# Verificar tipos de variables flotantes
print(f"Tipo de voltaje: {type(voltaje)}")
print(f"Tipo de constante_planck: {type(constante_planck)}")

Tipo de voltaje: <class 'float'>
Tipo de constante_planck: <class 'float'>


In [9]:
# Demostración de precisión de flotantes
print(f"0.1 + 0.2 = {0.1 + 0.2}")  # Nota: resultado no exacto por representación binaria
print(f"¿Es 0.1 + 0.2 == 0.3? {0.1 + 0.2 == 0.3}")
print("💡 Este comportamiento ilustra las limitaciones de precisión de los flotantes.")

0.1 + 0.2 = 0.30000000000000004
¿Es 0.1 + 0.2 == 0.3? False
💡 Este comportamiento ilustra las limitaciones de precisión de los flotantes.


### ✅❌ Booleanos (`bool`)

Los **booleanos** representan valores lógicos con dos estados posibles: `True` (verdadero) y `False` (falso). Constituyen la base para expresar condiciones y estados binarios.

#### Características principales:
- Dos valores posibles: `True` y `False`
- Representan estados o resultados de comparaciones
- Útiles para indicar activación/inactivación, aprobación/denegación, etc.
- En Python, `True` se asocia a 1 y `False` a 0, aunque su uso es eminentemente lógico

#### Ejemplos de uso:
- Estado de un dispositivo (encendido/apagado)
- Confirmación de acciones (aceptado/rechazado)
- Resultado de comparaciones (mayor/igual/menor)

#### Ejemplo en Python:
```python
es_mayor_edad = True
llueve = False
tiene_licencia = True
```

---

In [10]:
# Experimentando con booleanos
verdadero: bool = True
falso: bool = False
print(f"Valor verdadero: {verdadero}")

Valor verdadero: True


In [11]:
# Imprimir otros valores booleanos
print(f"Valor falso: {falso}")
print(f"Tipo de verdadero: {type(verdadero)}")

Valor falso: False
Tipo de verdadero: <class 'bool'>


In [12]:
# Booleanos como números
print(f"True como número: {True + 0}")
print(f"False como número: {False + 0}")
print(f"Suma True + True: {True + True}")

True como número: 1
False como número: 0
Suma True + True: 2


In [13]:
# Comparaciones que generan booleanos
temperatura_actual: float = 95.5
temperatura_maxima: float = 90.0
print(f"Temperatura actual: {temperatura_actual}°C")
print(f"Temperatura máxima: {temperatura_maxima}°C")

Temperatura actual: 95.5°C
Temperatura máxima: 90.0°C


### 📝 Cadenas (`str`)

Las **cadenas** son secuencias ordenadas e inmutables de caracteres. Se emplean para representar texto.

#### Características principales:
- Secuencias de caracteres Unicode (compatibles con tildes, emojis, etc.)
- Inmutables: su contenido no se modifica después de su creación
- Se definen con comillas simples `'...'`, dobles `"..."` o triples `'''...'''`
- Un carácter individual corresponde a una cadena de longitud 1

#### Ejemplos de uso:
- Nombres propios
- Mensajes en interfaces de usuario
- Códigos e identificadores
- Contenido de comunicaciones

#### Ejemplo en Python:
```python
letra = 'A'
nombre = "Ana"
mensaje = "Hola, mundo"
frase = "La programación es una disciplina rigurosa"
```

---

In [14]:
# Experimentando con cadenas
nombre_sensor: str = "Sensor de Temperatura DHT22"
id_componente: str = 'X-48-AB-v2'
letra: str = 'A'  # Un carácter es una cadena de longitud 1
cadena_vacia: str = ""
print(f"Nombre del sensor: {nombre_sensor}")

Nombre del sensor: Sensor de Temperatura DHT22


In [15]:
# Imprimir otros ejemplos de cadenas
print(f"ID del componente: {id_componente}")
print(f"Una letra: {letra}")
print(f"Cadena vacía: '{cadena_vacia}'")

ID del componente: X-48-AB-v2
Una letra: A
Cadena vacía: ''


In [16]:
# Verificar tipos y longitudes de cadenas
print(f"Tipo de nombre_sensor: {type(nombre_sensor)}")
print(f"Longitud del nombre: {len(nombre_sensor)} caracteres")
print(f"Longitud de una letra: {len(letra)} caracteres")

Tipo de nombre_sensor: <class 'str'>
Longitud del nombre: 27 caracteres
Longitud de una letra: 1 caracteres


In [17]:
# Operaciones con cadenas
print(f"Concatenación: {id_componente + '-' + letra}")
print(f"Repetición: {letra * 5}")
print(f"Mayúsculas: {nombre_sensor.upper()}")
print(f"Minúsculas: {id_componente.lower()}")

Concatenación: X-48-AB-v2-A
Repetición: AAAAA
Mayúsculas: SENSOR DE TEMPERATURA DHT22
Minúsculas: x-48-ab-v2


In [18]:
# Indexación y slicing de cadenas
print(f"Primer carácter: '{nombre_sensor[0]}'")
print(f"Último carácter: '{nombre_sensor[-1]}'")
print(f"Primeras 6 letras: '{nombre_sensor[:6]}'")
print(f"Palabra 'Sensor': '{nombre_sensor[0:6]}'")

Primer carácter: 'S'
Último carácter: '2'
Primeras 6 letras: 'Sensor'
Palabra 'Sensor': 'Sensor'


In [19]:
# Métodos útiles para ingeniería con cadenas
codigo_sensor: str = "TEMP_01_ACTIVO"
print(f"Código original: {codigo_sensor}")
print(f"¿Contiene 'TEMP'? {codigo_sensor.startswith('TEMP')}")
print(f"¿Está activo? {codigo_sensor.endswith('ACTIVO')}")
print(f"Separar por '_': {codigo_sensor.split('_')}" )

Código original: TEMP_01_ACTIVO
¿Contiene 'TEMP'? True
¿Está activo? True
Separar por '_': ['TEMP', '01', 'ACTIVO']


### 🔢 Enteros (`int`): métodos útiles

- `bit_length()`: cantidad de bits necesarios para representar el número en binario (sin signo).
- `to_bytes(length, byteorder)`: representación en bytes.
- `int.from_bytes(bytes, byteorder)`: conversión desde bytes a entero (método de clase).

In [44]:
# Métodos de int
n = 1023
print(f"n = {n}")
print(f"bit_length → {n.bit_length()} bits")

# Representación en bytes (2 bytes, big-endian)
b = n.to_bytes(2, byteorder='big')
print(f"to_bytes(2,'big') → {b}")
print(f"int.from_bytes(b,'big') → {int.from_bytes(b, byteorder='big')}")

n = 1023
bit_length → 10 bits
to_bytes(2,'big') → b'\x03\xff'
int.from_bytes(b,'big') → 1023


### 🌊 Flotantes (`float`): métodos útiles

- `as_integer_ratio()`: par `(numerador, denominador)` exacto de la fracción que representa el flotante.
- `is_integer()`: indica si el valor es un entero matemático.
- `hex()`: representación hexadecimal del flotante.
- `float.fromhex(cadena)`: construcción desde representación hexadecimal.

In [45]:
# Métodos de float
x = 3.5
y = 2.0
print(f"x = {x}")
print(f"x.as_integer_ratio() → {x.as_integer_ratio()}")
print(f"y = {y}, y.is_integer() → {y.is_integer()}")
print(f"x.hex() → {x.hex()}")
print(f"float.fromhex(x.hex()) → {float.fromhex(x.hex())}")

x = 3.5
x.as_integer_ratio() → (7, 2)
y = 2.0, y.is_integer() → True
x.hex() → 0x1.c000000000000p+1
float.fromhex(x.hex()) → 3.5


### 📝 Cadenas (`str`): métodos frecuentes

- Cambio de formato: `lower()`, `upper()`, `title()`, `capitalize()`.
- Limpieza: `strip()`, `lstrip()`, `rstrip()` (eliminan espacios en extremos).
- Búsqueda y reemplazo: `find()`, `rfind()`, `replace()`, `startswith()`, `endswith()`.
- División y unión: `split(sep)`, `join(iterable)`.
- Comprobaciones: `isalpha()`, `isdigit()`, `isalnum()`, `isnumeric()`.
- Formato: `format(...)` (alternativa a f-strings).

In [46]:
# Métodos de str
s = "  Ingeniería de Datos 2025 🚀  "
print(f"Original: '{s}'")
print(f"strip(): '{s.strip()}'")
print(f"lower(): '{s.lower()}'")
print(f"upper(): '{s.upper()}'")
print(f"title(): '{s.title()}'")

texto = "temperatura,sensor,ambiente"
partes = texto.split(',')
print(f"split(',') → {partes}")
print(f"'-'.join(partes) → {'-'.join(partes)}")

print(f"replace('sensor','actuador') → {texto.replace('sensor','actuador')}")
print(f"startswith('tem') → {texto.startswith('tem')}")
print(f"endswith('ente') → {texto.endswith('ente')}")

codigo = "X123"
print(f"{codigo}.isalnum() → {codigo.isalnum()}")
print(f"'123'.isdigit() → {'123'.isdigit()}")

Original: '  Ingeniería de Datos 2025 🚀  '
strip(): 'Ingeniería de Datos 2025 🚀'
lower(): '  ingeniería de datos 2025 🚀  '
upper(): '  INGENIERÍA DE DATOS 2025 🚀  '
title(): '  Ingeniería De Datos 2025 🚀  '
split(',') → ['temperatura', 'sensor', 'ambiente']
'-'.join(partes) → temperatura-sensor-ambiente
replace('sensor','actuador') → temperatura,actuador,ambiente
startswith('tem') → True
endswith('ente') → True
X123.isalnum() → True
'123'.isdigit() → True


### ✅❌ Booleanos (`bool`): consideraciones

- `bool` es una subclase de `int` con dos valores: `True` y `False`.
- No introduce métodos propios relevantes; el énfasis está en:
  - Conversión lógica: `bool(x)` determina la veracidad de `x` (truthiness).
  - Operadores lógicos: `and`, `or`, `not`.
  - Uso en condiciones, filtros y validaciones.

In [42]:
# Operaciones y conversiones con bool
valores = [0, 1, -1, "", "texto", [], [1], None]
print({repr(v): bool(v) for v in valores})

a, b = True, False
print(f"a and b → {a and b}")
print(f"a or b  → {a or b}")
print(f"not a    → {not a}")

# bool hereda de int (uso ilustrativo)
print(f"isinstance(True, int) → {isinstance(True, int)}")

{'0': False, '1': True, '-1': True, "''": False, "'texto'": True, '[]': False, '[1]': True, 'None': False}
a and b → False
a or b  → True
not a    → False
isinstance(True, int) → True


### ∅ `None` (tipo `NoneType`)

`None` representa la ausencia de valor. Es un único objeto singleton que indica “no hay dato”.

- Uso típico: valores por omisión, retorno de funciones sin resultado explícito, marcadores de ausencia.
- Comparación: utilizar `is None` y `is not None` (no `==`).
- Comportamiento lógico: `bool(None)` es `False`.

In [43]:
# Uso de None
resultado = None
print(f"resultado: {resultado}, type: {type(resultado)}")

# Patrón común: comprobar ausencia antes de asignar/procesar
if resultado is None:
    resultado = 0  # valor por defecto
print(f"resultado (tras inicialización): {resultado}")

# Comparaciones y veracidad
print(f"resultado is None → {resultado is None}")
print(f"bool(None) → {bool(None)}")

resultado: None, type: <class 'NoneType'>
resultado (tras inicialización): 0
resultado is None → False
bool(None) → False


---

## 1.1 Métodos útiles integrados de los tipos básicos

Esta sección presenta métodos y operaciones integradas que resultan útiles al trabajar con `int`, `float`, `str` y `bool`, además del valor especial `None`.

> Los métodos son funciones asociadas a un tipo/objeto que facilitan operaciones frecuentes.

---

## 2. Variables: nomenclatura, inicialización, asignación

---

Las **variables** almacenan referencias a valores en memoria. Funcionan como identificadores que permiten acceder y manipular dichos valores de manera controlada.

A continuación se presentan reglas de nomenclatura y operaciones básicas de inicialización, asignación y reasignación.

### Nomenclatura de variables
En Python, los nombres de variables deben cumplir:
- Iniciar con una letra o guión bajo (`_`).
- Contener letras, números y guión bajo (sin espacios ni tildes).
- Respetar mayúsculas/minúsculas (`edad` ≠ `Edad`).
- No coincidir con palabras reservadas (`if`, `for`, `class`, etc.).

Ejemplos válidos e inválidos:

In [ ]:
# Nombres válidos
nombre: str = 'Ana'
edad_usuario: int = 25
_saldo_cuenta: float = 100.0
temperatura_ambiente: float = 22.5

# Nombres inválidos (generarían error si se descomentan)
# 2nombres = 'Juan'   # No puede iniciar con número
# mi-variable = 10    # No puede contener guiones
# class = 5           # No puede ser palabra reservada

### Inicialización de variables
La inicialización consiste en asignar un valor por primera vez a una variable, creando su referencia en memoria.

In [ ]:
contador: int = 0  # Inicialización: se crea la variable y se le da un valor inicial

### Asignación de variables
La asignación establece o actualiza el valor asociado a una variable mediante el operador `=`.

In [ ]:
contador: int = 5  # Asignación: se cambia el valor de la variable

### Reasignación de variables
La reasignación modifica el valor previamente asociado a una variable. La identidad del nombre se mantiene; el contenido referenciado puede cambiar.

In [ ]:
contador: int = 10  # Reasignación: se vuelve a cambiar el valor

### 💻 Ejercicio: Intercambio de valores entre sensores

Contexto: En un sistema de control industrial, se dispone de dos sensores:
- `sensor_temperatura`: mide la temperatura del sistema.
- `sensor_presion`: mide la presión del sistema.

```python
sensor_temperatura = 23
sensor_presion = 39
```

Debido a un error de cableado, las lecturas están intercambiadas. Realice el intercambio correcto de valores entre las variables.

Condición: No introducir valores literales. Utilizar únicamente asignaciones (y variables auxiliares si es necesario).

---

In [47]:
sensor_temperatura: int = 23
sensor_presion: int = 39

print(f"Temperatura del sensor: {sensor_temperatura} °C")
print(f"Presión del sensor: {sensor_presion} hPa")

# Intercambio
variable_temporal: int = sensor_temperatura
sensor_temperatura, sensor_presion = sensor_presion, variable_temporal

print(f"Temperatura del sensor después del intercambio: {sensor_temperatura} °C")
print(f"Presión del sensor después del intercambio: {sensor_presion} hPa")

Temperatura del sensor: 23 °C
Presión del sensor: 39 hPa
Temperatura del sensor después del intercambio: 39 °C
Presión del sensor después del intercambio: 23 hPa


### 🔍 Verificación de tipos (type checking)

Es fundamental conocer el tipo de dato que se manipula para depurar, validar y documentar el código.

#### Objetivos
- Depuración: detectar errores relacionados con tipos
- Validación: comprobar que los datos cumplen el tipo esperado
- Documentación: facilitar la comprensión del código
- Robustez: prevenir fallos en tiempo de ejecución

#### Herramientas
- `type()`: retorna el tipo del objeto
- `isinstance()`: verifica pertenencia a un tipo/clase
- `dir()`: lista métodos y atributos disponibles

---

In [ ]:
# EXPLORACIÓN DE TIPOS DE DATOS
# Creamos variables de diferentes tipos
nombre: str = "Ana"              # string (cadena de caracteres)
edad: int = 20                   # int (entero)
altura: float = 1.65             # float (flotante)
es_estudiante: bool = True       # bool (booleano)
print(f"Nombre: {nombre}")
print(f"Tipo de nombre: {type(nombre)}")

Nombre: Ana
Tipo de nombre: <class 'str'>


In [31]:
# Imprimir edad y su tipo
print(f"Edad: {edad}")
print(f"Tipo de edad: {type(edad)}")

Edad: 20
Tipo de edad: <class 'int'>


In [32]:
# Imprimir altura y su tipo
print(f"Altura: {altura}")
print(f"Tipo de altura: {type(altura)}")

Altura: 1.65
Tipo de altura: <class 'float'>


In [33]:
# Imprimir si es estudiante y su tipo
print(f"Es estudiante: {es_estudiante}")
print(f"Tipo de es_estudiante: {type(es_estudiante)}")

Es estudiante: True
Tipo de es_estudiante: <class 'bool'>


In [34]:
# Verificaciones simples con type()
print(f"¿La edad es un entero? {type(edad) == int}")
print(f"¿El nombre es una cadena? {type(nombre) == str}")
print(f"¿La altura es un flotante? {type(altura) == float}")
print(f"¿es_estudiante es booleano? {type(es_estudiante) == bool}")

¿La edad es un entero? True
¿El nombre es una cadena? True
¿La altura es un flotante? True
¿es_estudiante es booleano? True


---

## 3. Conversión de tipos

---

En numerosas situaciones es necesario transformar datos entre tipos para realizar operaciones o presentar resultados correctamente.

### Motivación
- Convertir texto a números para cálculos
- Formatear números como texto para salida
- Validar y limpiar datos de entrada

#### Funciones principales
- `int()`: convierte a entero
- `float()`: convierte a flotante
- `str()`: convierte a cadena
- `bool()`: convierte a booleano

#### Ejemplo en Python
```python
edad_texto = "25"
edad = int(edad_texto)
altura_texto = "1.75"
altura = float(altura_texto)
nombre = str(12345)
activo = bool("algo")
```

Nota: La conversión requiere compatibilidad entre el valor origen y el tipo destino; de lo contrario, se produce un error.

El proceso de conversión de un tipo a otro se denomina casting.

---

In [ ]:
# CONVERSIÓN DE TIPOS
# Datos de ejemplo
numero_como_texto: str = "25"        # String que representa un entero
decimal_como_texto: str = "3.14"     # String que representa un flotante
numero_entero: int = 42              # Entero que queremos como string
numero_decimal: float = 2.71         # Flotante que queremos como entero

print("DATOS ORIGINALES:")
print(f"numero_como_texto: '{numero_como_texto}' (tipo: {type(numero_como_texto).__name__})")
print(f"decimal_como_texto: '{decimal_como_texto}' (tipo: {type(decimal_como_texto).__name__})")
print(f"numero_entero: {numero_entero} (tipo: {type(numero_entero).__name__})")
print(f"numero_decimal: {numero_decimal} (tipo: {type(numero_decimal).__name__})")

DATOS ORIGINALES:
numero_como_texto: '25' (tipo: str)
decimal_como_texto: '3.14' (tipo: str)
numero_entero: 42 (tipo: int)
numero_decimal: 2.71 (tipo: float)


In [ ]:
# String a número
numero: int = int(numero_como_texto)
print(f"String a int: '{numero_como_texto}' → {numero}")
decimal: float = float(decimal_como_texto)
print(f"String a float: '{decimal_como_texto}' → {decimal}")

String a int: '25' → 25
String a float: '3.14' → 3.14


In [ ]:
# Número a string y float a int
texto: str = str(numero_entero)
print(f"Int a string: {numero_entero} → '{texto}'")
entero: int = int(numero_decimal)
print(f"Float a int: {numero_decimal} → {entero}")

Int a string: 42 → '42'
Float a int: 2.71 → 2


In [38]:
# Conversiones a booleano
print(f"bool(1): {bool(1)}")
print(f"bool(0): {bool(0)}")
print(f"bool('hola'): {bool('hola')}")
print(f"bool(''): {bool('')}")

bool(1): True
bool(0): False
bool('hola'): True
bool(''): False


In [ ]:
# ⚠️ Conversiones que pueden dar error
try:
    numero_malo: int = int("hola")
except ValueError as e:
    print(f"Error al convertir 'hola' a int: {e}")

ValueError: invalid literal for int() with base 10: 'hola'

In [ ]:
# ⚠️ Conversiones que pueden dar error
try:
    decimal_malo: float = float("abc")
except ValueError as e:
    print(f"Error al convertir 'abc' a float: {e}")

---

## 4. Mutabilidad e inmutabilidad

---

En Python, algunos tipos permiten modificar su contenido después de creados (**mutables**) y otros no (**inmutables**). Este comportamiento afecta la gestión de memoria y la forma en que se comparten referencias entre variables.

### Definiciones
- **Mutables:** su contenido puede modificarse sin cambiar la identidad del objeto (ej.: `list`, `dict`, `set`).
- **Inmutables:** su contenido no puede alterarse; toda modificación genera un objeto nuevo (ej.: `int`, `float`, `bool`, `str`).

#### Tipos inmutables
- `int`, `float`, `bool`, `str`

#### Tipos mutables
- `list`, `dict`, `set`

#### Ejemplos
```python
# Inmutabilidad
a = 5
a = 6  # se crea un nuevo objeto

# Mutabilidad
lista = [1, 2, 3]
lista.append(4)  # se modifica el mismo objeto
```

Importancia: Al compartir objetos mutables entre variables, los cambios se reflejan en todas las referencias. Con objetos inmutables, cada reasignación crea una nueva referencia.

---

In [ ]:
# DEMOSTRACIÓN BÁSICA DE MUTABILIDAD
# Ejemplo con tipos INMUTABLES (números y strings)
numero: int = 42
print(f"Número original: {numero}")
numero = numero + 1
print(f"Número después de +1: {numero}")
print("💡 Nota: Se crea un nuevo número; el original no cambia.")

In [ ]:
# Strings también son inmutables
texto: str = "Hola"
print(f"Texto original: '{texto}'")
texto = texto + " mundo"
print(f"Texto después de concatenar: '{texto}'")
print("💡 Nota: Se crea un nuevo string; el original no cambia.")

In [ ]:
# Ejemplo con listas (MUTABLES)
lista_numeros: list[int] = [1, 2, 3]
print(f"Lista original: {lista_numeros}")
lista_numeros.append(4)
print(f"Lista después de agregar 4: {lista_numeros}")
print("💡 Nota: Se modifica el mismo objeto (lista).")

In [ ]:
# Demostración con variables que apuntan a la misma lista
lista_a: list[int] = [1, 2, 3]
lista_b: list[int] = lista_a  # Ambas variables referencian la misma lista
print(f"lista_a: {lista_a}")
print(f"lista_b: {lista_b}")
lista_b.append(4)
print("Después de lista_b.append(4):")
print(f"lista_a: {lista_a}")  # También cambia
print(f"lista_b: {lista_b}")
print("💡 Nota: Al compartir referencias a objetos mutables, los cambios se reflejan en todas las variables.")

In [ ]:
# Resumen de mutabilidad
print("• int, float, bool, str → INMUTABLES")
print("• list, dict → MUTABLES")
print("• Atención: al compartir listas entre variables, los cambios afectan a todas las referencias")

## 🎯 Resumen y ejercicios de repaso

Se presentó una síntesis de los tipos de datos básicos y el manejo de variables en Python.

### 📚 Contenidos revisados

1. **Tipos de datos básicos**:
   - `int`: números enteros para conteos e identificadores
   - `float`: números decimales para mediciones
   - `bool`: valores lógicos para condiciones
   - `str`: texto para nombres, mensajes e identificadores

2. **Variables**:
   - Contenedores que almacenan referencias a objetos
   - Nombres descriptivos mejoran la legibilidad
   - Convención de nomenclatura: snake_case

3. **Conversión de tipos**:
   - `int()`, `float()`, `str()`, `bool()` para transformar datos
   - Manejo de errores en conversiones

4. **Mutabilidad**:
   - Inmutables: `int`, `float`, `bool`, `str`
   - Mutables: `list`, `dict`, `set`
   - Atención con referencias compartidas

---

## 📝 Ejercicios de práctica

A continuación se proponen ejercicios organizados por tema para consolidar los conceptos.

### 1️⃣ Ejercicios: Tipos de datos fundamentales

**Ejercicio 1.1 — Enteros**
```python
# Declarar variables para:
# - La edad
# - El año actual
# - La cantidad de días en una semana
# - Un entero negativo que represente temperatura bajo cero
```

**Ejercicio 1.2 — Flotantes**
```python
# Declarar variables para:
# - Altura en metros
# - Valor de pi (3.14159)
# - Temperatura corporal (36.5)
# - Un número pequeño en notación científica (1.5e-10)
```

**Ejercicio 1.3 — Booleanos**
```python
# Declarar variables booleanas para:
# - Licencia de conducir
# - Si está lloviendo hoy
# - Si 5 es mayor que 3
# - Si la edad es mayor o igual a 18
```

**Ejercicio 1.4 — Cadenas**
```python
# Declarar variables string para:
# - Nombre completo
# - Color favorito
# - Frase motivacional
# - Cadena vacía
```

### 2️⃣ Ejercicios: Variables (nomenclatura, inicialización, asignación)

**Ejercicio 2.1 — Nomenclatura correcta**
```python
# Indicar cuáles nombres son válidos en Python (✅) o inválidos (❌):
# mi_edad →
# 2nombres →
# nombre_completo →
# class →
# _temperatura →
# mi-variable →
# nombre2 →
# Nombre →
```

**Ejercicio 2.2 — Asignación múltiple**
```python
# Declarar en una sola línea:
# nombre = "Juan", edad = 25, altura = 1.75
```

**Ejercicio 2.3 — Reasignación**
```python
# Partir de: contador = 0
# Incrementar en 5 y luego duplicar el valor
# Determinar el valor final
```

### 3️⃣ Ejercicios: Conversión de tipos

**Ejercicio 3.1 — Conversiones básicas**
```python
# Convertir los siguientes valores:
texto_edad = "25"        # → a int con int()
texto_precio = "19.99"   # → a float con float()
numero = 42              # → a str con str()
numero_decimal = 3.7     # → a int con int()
```

**Ejercicio 3.2 — Conversiones a booleano**
```python
# Convertir a booleano con bool():
numero_uno = 1
numero_cero = 0
texto_vacio = ""
texto_lleno = "hola"
```

### 4️⃣ Ejercicios: Mutabilidad e inmutabilidad

**Ejercicio 4.1 — Clasificación**
```python
# Clasificar como MUTABLE o INMUTABLE:
int  →
str  →
list →
float →
bool →
```

**Ejercicio 4.2 — Comportamiento inmutable**
```python
texto = "Hola"
texto_nuevo = texto + " mundo"
print(texto)        # ¿Cambia el texto original?
print(texto_nuevo)  # ¿Qué contiene texto_nuevo?
```

**Ejercicio 4.3 — Comportamiento mutable**
```python
lista_a = [1, 2, 3]
lista_b = lista_a
lista_b.append(4)
print(lista_a)  # ¿Contenido de lista_a?
print(lista_b)  # ¿Contenido de lista_b?
```

### 5️⃣ Ejercicios integrados

**Ejercicio 5.1 — Datos personales básicos**
```python
# Crear variables para:
# - nombre (string)
# - edad_texto (string "25")
# - edad_numero (convierte edad_texto a int)
# - altura_texto (string "1.75")
# - altura_numero (convierte altura_texto a float)
```

**Ejercicio 5.2 — Verificación de tipos**
```python
# Dadas estas variables:
a = 10
b = 3.14
c = "Python"
d = True

# Verificar el tipo con type():
# print(type(a))
# print(type(b))
# print(type(c))
# print(type(d))
```

### 6️⃣ Ejercicios de repaso

**Ejercicio 6.1 — Intercambio de variables**
```python
# Intercambiar los valores usando una variable temporal:
x = 10
y = 20
# 1. temporal = x
# 2. x = y
# 3. y = temporal
```

**Ejercicio 6.2 — Práctica de conversiones**
```python
numero_como_texto = "123"
decimal_como_texto = "45.67"

# Convertir numero_como_texto a int
# Convertir decimal_como_texto a float
# Convertir el número 89 a string
# Convertir el string "89" a int
```

---

### 📋 Instrucciones para resolver

1. Copiar cada ejercicio a una nueva celda de código.
2. Resolver paso a paso.
3. Ejecutar para verificar resultados.
4. Experimentar modificando valores.
5. Consultar dudas cuando sea necesario.